### PR TITLE
New IE for RTVSLO (rtvslo.si)

### PR DESCRIPTION
**Pull request by iw0nderhow**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This PR adds an extractor for the VOD platform of RTVSLO, the Slovenian public broadcaster.

I sadly can't test whether the extractor downloads georestricted content successfully as I do not have access to a Slovenian IP.

RTVSLO also provides paywalled content: I also was not able to test that due to the same reason. 
There is a way to determine, however, *that* the content is paywalled: should I add detection for it?

### CODE REVIEW

**Comment by iw0nderhow** [comment]
Just realized I can get rid of a lot of this by just using `_extract_wowza_formats`.
Also, the `mediaFiles_sl` and `addaptiveMedia_sl` formats the API returns are sign language.
Is there a convention for sign language formats or do I just return them like normal and hope for the best?

**Comment by iw0nderhow** [comment]
I think this is ready for review (albeit the subtitles extraction does not spark joy in me.)
One more question: the website sometimes returns an English audio track in its DASH manifest which obviously isn't English. Should I replace the language tag or just pass through the erroneous data?

File: yt_dlp/extractor/rtvslo.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+from .common import InfoExtractor
+from ..utils import ExtractorError, traverse_obj, parse_duration, unified_timestamp
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:2]: split line

File: yt_dlp/extractor/rtvslo.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        }, {
+            'url': 'https://4d.rtvslo.si/arhiv/dnevnik/174842550',
+            'only_matching': True
+        }
+    ]
+
+    def _real_extract(self, url):
+        v_id = self._match_id(url)
+        meta = self._download_json(self._API_BASE.format('getRecordingDrm', v_id), v_id).get('response')
+        date = unified_timestamp(meta.get('broadcastDate') or meta.get('broadcastDates')[0])
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:69]: traverse_obj

File: yt_dlp/extractor/rtvslo.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        thumbs = [{'id': k, 'url': v} for (k, v) in meta.get('images').items()]
+        SUB_LANGS_MAP = {'Slovenski': 'sl', }
+
+        subs = {}
+        for s in traverse_obj(meta, 'subs', 'subtitles', default={}):
+            if s.get('language') in SUB_LANGS_MAP.keys():
+                s['language'] = SUB_LANGS_MAP[s['language']]
+            if not subs.get(s.get('language'), False):
+                subs[s.get('language')] = []
+            subs[s.get('language')].append({'url': s.get('file'), 'format': s.get('format').lower()})
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:80]:
```suggestion
            lang = SUB_LANGS_MAP.get(s.get('language'), s.get('language') or 'und'))
            subs.setdefault(lang, []).append({'url': s.get('file'), 'format': s.get('format').lower()})
```

File: yt_dlp/extractor/rtvslo.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+        }, {
+            'url': 'https://4d.rtvslo.si/arhiv/dnevnik/174842550',
+            'only_matching': True
+        }
+    ]
+
+    def _real_extract(self, url):
+        v_id = self._match_id(url)
+        meta = self._download_json(self._API_BASE.format('getRecordingDrm', v_id), v_id).get('response')
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:68]: Is this supposed to be fatal or not?
```suggestion
        meta = self._download_json(self._API_BASE.format('getRecordingDrm', v_id), v_id)['response']
```
OR

```suggestion
        meta = self._download_json(self._API_BASE.format('getRecordingDrm', v_id), v_id, fatal=False).get('response') or []
```

File: yt_dlp/extractor/rtvslo.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                s['language'] = SUB_LANGS_MAP[s['language']]
+            if not subs.get(s.get('language'), False):
+                subs[s.get('language')] = []
+            subs[s.get('language')].append({'url': s.get('file'), 'format': s.get('format').lower()})
+
+        jwt = meta.get('jwt')
+        if not jwt:
+            raise ExtractorError('Site did not provide an authentication token, cannot proceed.')
+
+        media = self._download_json(self._API_BASE.format('getMedia', v_id) + f'&jwt={jwt}', v_id).get('response')
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:86]:
```suggestion
        media = self._download_json(self._API_BASE.format('getMedia', v_id), v_id, query={jwt: jwt}).get('response')
```

Comment (pukkandan) [reply yt_dlp/extractor/rtvslo.py:86]: Similar to above, decide whether this should be fatal or not! Just using `.get` without other precautions don't accomplish anything

File: yt_dlp/extractor/rtvslo.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+        formats = []
+        if media.get('addaptiveMedia', False):
+            formats = self._extract_wowza_formats(
+                traverse_obj(media, ('addaptiveMedia', 'hls_sec'), expected_type=str),
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:91]:
```suggestion
                traverse_obj(media, ('addaptiveMedia', 'hls_sec'), expected_type=url_or_none),
```

File: yt_dlp/extractor/rtvslo.py (hunk 3 above)
Comment (pukkandan) [reply yt_dlp/extractor/rtvslo.py:80]: `.get.lower` is fatal

File: yt_dlp/extractor/rtvslo.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                if traverse_obj(f, ('streams', strm)):
+                    formats.append({
+                        'bitrate': f.get('bitrate'),
+                        'url': traverse_obj(f, ('streams', strm)),
+                        'filesize': f.get('filesize'),
+                        'width': f.get('width'),
+                        'height': f.get('height'),
+                        'ext': f.get('mediaType').lower(),
+                        'format_id': f'files_{strm}_{f.get("mediaType").lower()}_{f.get("bitrate")}'
+                    })
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:104]:
```suggestion
            formats.extend({
                'bitrate': f.get('bitrate'),
                'url': traverse_obj(f, ('streams', strm)),
                'filesize': f.get('filesize'),
                'width': f.get('width'),
                'height': f.get('height'),
                'ext': f.get('mediaType').lower(),
                'format_id': f'files_{strm}_{f.get("mediaType").lower()}_{f.get("bitrate")}'
            } for f in media.get('mediaFiles') if traverse_obj(f, ('streams', strm)))
```

(don't need to do this if making everything non-fatal over-complicates the comprehension)

Comment (pukkandan) [reply yt_dlp/extractor/rtvslo.py:104]: `for in .get` and `.get.lower` are fatal

File: yt_dlp/extractor/rtvslo.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+        if media.get('addaptiveMedia_sl', False):
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:106]:
```py
url = ...
if url:
```

File: yt_dlp/extractor/rtvslo.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            for f in self._extract_wowza_formats(
+                traverse_obj(media, ('addaptiveMedia_sl', 'hls_sec')), v_id, skip_protocols=['smil']
+            ):
+                f.update({'format_note': 'Sign language interpretation', 'preference': -3})
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:110]: -10 is convention

File: yt_dlp/extractor/rtvslo.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+                        'bitrate': f.get('bitrate'),
+                        'url': traverse_obj(f, ('streams', strm)),
+                        'filesize': f.get('filesize'),
+                        'width': f.get('width'),
+                        'height': f.get('height'),
+                        'ext': f.get('mediaType').lower(),
+                        'format_id': f'files-sl_{strm}_{f.get("mediaType").lower()}_{f.get("bitrate")}',
+                        'format_note': 'Sign language interpretation',
+                        'preference': -3
+                    })
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:125]: `.extend`, make sure everything is non-fatal

File: yt_dlp/extractor/rtvslo.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+        self._sort_formats(formats)
+
+        if any('intermission.mp4' in x.get('url', '') for x in formats):
+            self.raise_geo_restricted(countries=self._GEO_COUNTRIES, metadata_available=True)
+        if any('dummy_720p.mp4' in x.get('manifest_url', '') for x in formats) and meta.get('stub', '') == 'error':
+            raise ExtractorError(f'{self.IE_NAME} said: Clip not available')
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:132]:
```suggestion
            raise ExtractorError(f'{self.IE_NAME} said: Clip not available', expected=True)
```

File: yt_dlp/extractor/rtvslo.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+            'thumbnails': thumbs,
+            'subtitles': subs,
+            'title': meta.get('title'),
+            'id': v_id,
+            'description': meta.get('description'),
+            'formats': formats,
+            'timestamp': date,
+            'release_timestamp': unified_timestamp(meta.get('recordingDate')),
+            'duration': meta.get('duration') or parse_duration(meta.get('length')),
+            'webpage_url': ''.join(traverse_obj(meta, ('canonical', ('domain', 'path')))),
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:144]: needed?

File: yt_dlp/extractor/rtvslo.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+                    })
+
+        self._sort_formats(formats)
+
+        if any('intermission.mp4' in x.get('url', '') for x in formats):
+            self.raise_geo_restricted(countries=self._GEO_COUNTRIES, metadata_available=True)
+        if any('dummy_720p.mp4' in x.get('manifest_url', '') for x in formats) and meta.get('stub', '') == 'error':
+            raise ExtractorError(f'{self.IE_NAME} said: Clip not available')
+
+        info = {
```
Comment (pukkandan) [inline yt_dlp/extractor/rtvslo.py:134]:
```suggestion
        return {
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
> One more question: the website sometimes returns an English audio track in its DASH manifest which obviously isn't English. Should I replace the language tag or just pass through the erroneous data?

Replace if you can

**Comment by iw0nderhow** [comment]
Is it normal for the tests to use a different format ordering than the normal yt_dlp invocation? 

My tests are failing because the tests (`python3 test/test_download.py TestDownload.test_RTVSLO_all`) choose the HDS stream while just running `python3 -m yt_dlp ...` chooses two DASH formats 🤔

**Comment by pukkandan** [comment]
Yes, tests use the format selector `best/bestvideo`

**Comment by iw0nderhow** [comment]
So, there's just one problem left (mentioned it on Discord): For the thumbnail download to succeed, I need to pass custom headers. The `Accept: text/html,...` header in `std_headers` makes the CDN return a HTML page with the image embedded instead of the actual image. :/

File: yt_dlp/extractor/rtvslo.py (hunk 12 above)
Comment (iw0nderhow) [reply yt_dlp/extractor/rtvslo.py:143]: GitHub highlights seem to be slightly broken here, I'm seeing different code sections highlighted between the email and the web interface.
Concerning `intermission.mp4` and `dummy_720p.mp4`: if I don't filter for them, an error video will be downloaded. If the user doesn't check the returned video, they will believe the video was downloaded successfully and be surprised once they actually view it (which may be after the video is unpublished).

Comment (iw0nderhow) [reply yt_dlp/extractor/rtvslo.py:143]: Oh my, now GitHub's finally working. 

Reassembling `webpage_url` like this is probably expendable - however, accessing a link to a video on `4d.rtvslo.si` currently returns a redirect; and judging by its main page, the portal may be sunset some time in the future. So, I assemble a link to the current portal (`365.rtvslo.si`) from the data I receive.

**Comment by pukkandan** [comment]
297e9952b65cfd03aab2d0b97c8cf18243d12578 allows `http_headers` to be defined for thumbnails

**Comment by pukkandan** [comment]
Sorry for the delay. This PR kinda got lost among the others.

I have pushed some (mostly cosmetic) changes. Go though and pls confirm everything is fine

**Comment by iw0nderhow** [comment]
The changes look fine, just had to think about the removal of `mediaFiles_sl` processing. In my tests, the array has always been empty though. We can probably omit it.

**Comment by dmovztoq** [comment]
Hi

I just noticed this and thank you very much, been waiting for this for quite a while. There's been an attempt back then here https://github.com/ytdl-org/youtube-dl/pull/28134 but you might have already seen that by now, either way it may not help, the 4D platform got updated a bit to "356" with some changes, but the old 4D links seem to keep working.

So far so good, I did a quick test with dash video and audio and it plays fine in VLC and MPV.

I can help you test if there's anything you'd wish. You can give me a few links that you can't access.

BTW: There are also livestreams here, but that's a bonus if you manage to get that working, it's not that important for me personally tho. There are cases where they do not archive full transmissions, for one the sport transmissions (even when RTVSLO owns the licenses) aren't archived fully (only reports/excerpts) in their 356 VODs. And it would be cool if it could "start from beginning" too. Livestreams have a 2 hour rewind buffer. But I just realized that I think livestreams are most definitely geo-blocked, oh well.

**Pull request merged**